### DISCUSSION TITLE
What is the advantage of the `useMutation` callbacks over the native Promise syntax?

### DISCUSSION THREAD

**Discussion by githorse** [TanStack/query#5807] [Q&A]
Why do we need callbacks on `useMutation`? Is there a distinction between 

```
const {mutate} = useMutation({
  mutationFn: cat => post(`/cats`, cat),
  onSuccess: cat => toast.success('Made a cat')
})
```

and

```
const {mutate} = useMutation({
  mutationFn: cat => post(`/cats`, cat)
    .then(cat => toast.success('Made a cat'))
})
```

Related to https://github.com/TanStack/query/discussions/2651, but there the difference comes down to the way promise rejections are handled, which (I don't think?) is relevant here.

**Comment by TkDodo** [comment]
you can use both, there isn't a big difference. The callbacks are nice to read, and they can be overwritten if you use an abstraction where the mutationFn is provided for you, like with `trpc`. Or, if you have a `defaultMutationFn`.

**Comment by githorse** [comment]
Worth noting one other distinction is that, with the native promise version, you need to be sure that any `.then()` returns the original data in order for the mutation function to make sense:

```
// ❌ bad
mutationFn: cat => post(`/cats`, cat).then(cat => toast.success('Made a cat'))

// ✔️ good
mutationFn: cat => post(`/cats`, cat).then(cat => {toast.success('Made a cat'); return cat})
```
